File: c_programs_and_containers/build_blocks/knowledge_base/mqtt/kv_reader/kv_store_reader.c
/**
 * kv_store_reader.c - Reads retained key/value messages from an MQTT broker.
 *
 * Translated from Python KVStoreReader class.
 * Requires: libmosquitto (mosquitto-dev)
 *   apt install libmosquitto-dev
 *   Link with: -lmosquitto -lpthread
 */

 #include "kv_store_reader.h"

 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <errno.h>
 #include <mosquitto.h>
 
 /* ── Internal helpers ───────────────────────────────────────────────── */
 
 static void kvr_now(struct timespec *ts)
 {
     clock_gettime(CLOCK_REALTIME, ts);
 }
 
 static void kvr_deadline(struct timespec *out, double offset_sec)
 {
     kvr_now(out);
     long whole = (long)offset_sec;
     long frac  = (long)((offset_sec - (double)whole) * 1e9);
     out->tv_nsec += frac;
     out->tv_sec  += whole + out->tv_nsec / 1000000000L;
     out->tv_nsec  = out->tv_nsec % 1000000000L;
 }
 
 /**
  * Check if `topic` is one of the configured sentinels.
  * Caller must hold sentinel_lock.
  */
 static bool kvr_is_sentinel(kvr_store_reader_t *kv, const char *topic)
 {
     for (int i = 0; i < kv->sentinel_count; i++) {
         if (strcmp(kv->sentinels[i], topic) == 0) return true;
     }
     return false;
 }
 
 /**
  * Add a KV entry. Overwrites if topic already exists.
  * Caller must hold entries_lock.
  */
 static void kvr_store_entry(kvr_store_reader_t *kv,
                             const char *topic, const char *value)
 {
     /* Check for existing topic — overwrite */
     for (int i = 0; i < kv->entry_count; i++) {
         if (kv->entries[i].active &&
             strcmp(kv->entries[i].topic, topic) == 0) {
             strncpy(kv->entries[i].value, value, KVR_MAX_VALUE_LEN - 1);
             kv->entries[i].value[KVR_MAX_VALUE_LEN - 1] = '\0';
             return;
         }
     }
     /* Append new entry */
     if (kv->entry_count < KVR_MAX_ENTRIES) {
         kvr_entry_t *e = &kv->entries[kv->entry_count++];
         strncpy(e->topic, topic, KVR_MAX_TOPIC_LEN - 1);
         e->topic[KVR_MAX_TOPIC_LEN - 1] = '\0';
         strncpy(e->value, value, KVR_MAX_VALUE_LEN - 1);
         e->value[KVR_MAX_VALUE_LEN - 1] = '\0';
         e->active = true;
     }
 }
 
 /* ── Mosquitto callbacks ────────────────────────────────────────────── */
 
 static void on_connect_cb(struct mosquitto *mosq, void *obj, int reason_code)
 {
     (void)mosq;
     kvr_store_reader_t *kv = (kvr_store_reader_t *)obj;
 
     pthread_mutex_lock(&kv->state_lock);
     kv->connected    = (reason_code == 0);
     kv->connect_done = true;
     pthread_cond_signal(&kv->connect_cond);
     pthread_mutex_unlock(&kv->state_lock);
 
     if (kv->connected) {
         printf("[Connected] Successfully connected to %s:%d\n",
                kv->config.host, kv->config.port);
     } else {
         printf("[Connection Failed] reason_code=%d\n", reason_code);
     }
 }
 
 static void on_disconnect_cb(struct mosquitto *mosq, void *obj, int reason_code)
 {
     (void)mosq;
     kvr_store_reader_t *kv = (kvr_store_reader_t *)obj;
 
     pthread_mutex_lock(&kv->state_lock);
     kv->connected = false;
     pthread_mutex_unlock(&kv->state_lock);
 
     printf("[Disconnected] reason_code=%d\n", reason_code);
 }
 
 static void on_message_cb(struct mosquitto *mosq, void *obj,
                           const struct mosquitto_message *msg)
 {
     (void)mosq;
     kvr_store_reader_t *kv = (kvr_store_reader_t *)obj;
 
     /* Decode payload to string */
     char payload[KVR_MAX_VALUE_LEN];
     if (msg->payload && msg->payloadlen > 0) {
         int copy_len = msg->payloadlen;
         if (copy_len >= KVR_MAX_VALUE_LEN) copy_len = KVR_MAX_VALUE_LEN - 1;
         memcpy(payload, msg->payload, copy_len);
         payload[copy_len] = '\0';
     } else {
         payload[0] = '\0';
     }
 
     /* Check if this is a sentinel topic */
     pthread_mutex_lock(&kv->sentinel_lock);
     if (kvr_is_sentinel(kv, msg->topic)) {
         kv->sentinel_fired = true;
         pthread_cond_signal(&kv->sentinel_cond);
         pthread_mutex_unlock(&kv->sentinel_lock);
         return;
     }
     pthread_mutex_unlock(&kv->sentinel_lock);
 
     /* Truncated preview for log */
     char preview[54];
     int plen = (int)strlen(payload);
     int clen = plen < 50 ? plen : 50;
     memcpy(preview, payload, clen);
     preview[clen] = '\0';
 
     /* Normal KV handling */
     pthread_mutex_lock(&kv->entries_lock);
     if (msg->retain) {
         kvr_store_entry(kv, msg->topic, payload);
         printf("[Retained] %s => %s%s\n", msg->topic, preview,
                plen > 50 ? "..." : "");
     } else {
         printf("[Non-retained] %s => %s%s\n", msg->topic, preview,
                plen > 50 ? "..." : "");
     }
     pthread_mutex_unlock(&kv->entries_lock);
 }
 
 static void on_subscribe_cb(struct mosquitto *mosq, void *obj,
                             int mid, int qos_count, const int *granted_qos)
 {
     (void)mosq;
     (void)qos_count;
     (void)granted_qos;
     kvr_store_reader_t *kv = (kvr_store_reader_t *)obj;
 
     printf("[Subscribed] mid=%d\n", mid);
 
     pthread_mutex_lock(&kv->sub_lock);
     kv->sub_acked = true;
     pthread_cond_signal(&kv->sub_cond);
     pthread_mutex_unlock(&kv->sub_lock);
 }
 
 /* ── Public API ─────────────────────────────────────────────────────── */
 
 void kvr_config_init(kvr_config_t *cfg)
 {
     memset(cfg, 0, sizeof(*cfg));
     strncpy(cfg->host, "localhost", sizeof(cfg->host) - 1);
     cfg->port            = 1883;
     strncpy(cfg->client_id, "kv-reader", sizeof(cfg->client_id) - 1);
     cfg->keepalive       = 60;
     cfg->use_mqttv5      = false;
     cfg->has_credentials = false;
     cfg->clean_session   = true;
 }
 
 int kvr_init(kvr_store_reader_t *kv, const kvr_config_t *cfg)
 {
     memset(kv, 0, sizeof(*kv));
     kv->config = *cfg;
 
     mosquitto_lib_init();
 
     kv->mosq = mosquitto_new(cfg->client_id, cfg->clean_session, kv);
     if (!kv->mosq) {
         fprintf(stderr, "[Error] mosquitto_new failed: %s\n", strerror(errno));
         return -1;
     }
 
     if (cfg->has_credentials) {
         mosquitto_username_pw_set(kv->mosq, cfg->username, cfg->password);
     }
 
     if (cfg->use_mqttv5) {
         int ver = MQTT_PROTOCOL_V5;
         mosquitto_int_option(kv->mosq, MOSQ_OPT_PROTOCOL_VERSION, ver);
     }
 
     mosquitto_connect_callback_set(kv->mosq, on_connect_cb);
     mosquitto_disconnect_callback_set(kv->mosq, on_disconnect_cb);
     mosquitto_message_callback_set(kv->mosq, on_message_cb);
     mosquitto_subscribe_callback_set(kv->mosq, on_subscribe_cb);
 
     pthread_mutex_init(&kv->entries_lock, NULL);
     pthread_mutex_init(&kv->sentinel_lock, NULL);
     pthread_cond_init(&kv->sentinel_cond, NULL);
     pthread_mutex_init(&kv->sub_lock, NULL);
     pthread_cond_init(&kv->sub_cond, NULL);
     pthread_mutex_init(&kv->state_lock, NULL);
     pthread_cond_init(&kv->connect_cond, NULL);
 
     return 0;
 }
 
 bool kvr_connect(kvr_store_reader_t *kv, double timeout_sec)
 {
     pthread_mutex_lock(&kv->state_lock);
     kv->connect_done = false;
     kv->running      = true;
     pthread_mutex_unlock(&kv->state_lock);
 
     int rc = mosquitto_connect(kv->mosq, kv->config.host,
                                kv->config.port, kv->config.keepalive);
     if (rc != MOSQ_ERR_SUCCESS) {
         fprintf(stderr, "[Error] Connection failed: %s\n", mosquitto_strerror(rc));
         kv->running = false;
         return false;
     }
 
     rc = mosquitto_loop_start(kv->mosq);
     if (rc != MOSQ_ERR_SUCCESS) {
         fprintf(stderr, "[Error] loop_start failed: %s\n", mosquitto_strerror(rc));
         kv->running = false;
         return false;
     }
 
     /* Wait for CONNACK */
     struct timespec deadline;
     kvr_deadline(&deadline, timeout_sec);
 
     pthread_mutex_lock(&kv->state_lock);
     while (!kv->connect_done) {
         int wait_rc = pthread_cond_timedwait(&kv->connect_cond,
                                              &kv->state_lock, &deadline);
         if (wait_rc == ETIMEDOUT) break;
     }
     bool ok = kv->connected;
     pthread_mutex_unlock(&kv->state_lock);
 
     if (!ok) {
         fprintf(stderr, "[Error] Connection timeout or refused (%s:%d)\n",
                 kv->config.host, kv->config.port);
         kv->running = false;
         mosquitto_loop_stop(kv->mosq, true);
         return false;
     }
 
     return true;
 }
 
 void kvr_disconnect(kvr_store_reader_t *kv)
 {
     kv->running = false;
     mosquitto_disconnect(kv->mosq);
     mosquitto_loop_stop(kv->mosq, false);
 
     pthread_mutex_lock(&kv->state_lock);
     kv->connected = false;
     pthread_mutex_unlock(&kv->state_lock);
 }
 
 void kvr_destroy(kvr_store_reader_t *kv)
 {
     if (kv->running || kv->connected) {
         kvr_disconnect(kv);
     }
     if (kv->mosq) {
         mosquitto_destroy(kv->mosq);
         kv->mosq = NULL;
     }
     pthread_mutex_destroy(&kv->entries_lock);
     pthread_mutex_destroy(&kv->sentinel_lock);
     pthread_cond_destroy(&kv->sentinel_cond);
     pthread_mutex_destroy(&kv->sub_lock);
     pthread_cond_destroy(&kv->sub_cond);
     pthread_mutex_destroy(&kv->state_lock);
     pthread_cond_destroy(&kv->connect_cond);
 
     mosquitto_lib_cleanup();
 }
 
 int kvr_read_pattern(kvr_store_reader_t *kv,
                      const char *pattern,
                      int qos,
                      double timeout_sec,
                      const char *sentinel_topics[],
                      bool wait_for_sentinel,
                      kvr_entry_t *out_entries,
                      int max_entries)
 {
     if (!kv->connected) {
         fprintf(stderr, "[Error] Not connected. Call kvr_connect() first.\n");
         return 0;
     }
 
     /* Clear previous entries */
     pthread_mutex_lock(&kv->entries_lock);
     kv->entry_count = 0;
     pthread_mutex_unlock(&kv->entries_lock);
 
     /* Setup sentinels */
     pthread_mutex_lock(&kv->sentinel_lock);
     kv->sentinel_fired = false;
     kv->sentinel_count = 0;
     if (sentinel_topics) {
         for (int i = 0; sentinel_topics[i] != NULL && i < KVR_MAX_SENTINELS; i++) {
             strncpy(kv->sentinels[i], sentinel_topics[i], KVR_MAX_TOPIC_LEN - 1);
             kv->sentinels[i][KVR_MAX_TOPIC_LEN - 1] = '\0';
             kv->sentinel_count++;
         }
     }
     pthread_mutex_unlock(&kv->sentinel_lock);
 
     /* Subscribe to pattern */
     pthread_mutex_lock(&kv->sub_lock);
     kv->sub_acked = false;
     pthread_mutex_unlock(&kv->sub_lock);
 
     int rc = mosquitto_subscribe(kv->mosq, NULL, pattern, qos);
     if (rc != MOSQ_ERR_SUCCESS) {
         fprintf(stderr, "[Error] Subscribe failed for %s: %s\n",
                 pattern, mosquitto_strerror(rc));
         /* Clear sentinels */
         pthread_mutex_lock(&kv->sentinel_lock);
         kv->sentinel_count = 0;
         pthread_mutex_unlock(&kv->sentinel_lock);
         return 0;
     }
 
     /* Wait for SUBACK (best-effort, 2 second cap) */
     {
         struct timespec sub_deadline;
         kvr_deadline(&sub_deadline, 2.0);
         pthread_mutex_lock(&kv->sub_lock);
         while (!kv->sub_acked) {
             int w = pthread_cond_timedwait(&kv->sub_cond, &kv->sub_lock,
                                            &sub_deadline);
             if (w == ETIMEDOUT) break;
         }
         pthread_mutex_unlock(&kv->sub_lock);
     }
 
     /* Wait for retained messages: sentinel or timeout */
     if (wait_for_sentinel && kv->sentinel_count > 0) {
         struct timespec deadline;
         kvr_deadline(&deadline, timeout_sec);
         pthread_mutex_lock(&kv->sentinel_lock);
         while (!kv->sentinel_fired) {
             int w = pthread_cond_timedwait(&kv->sentinel_cond,
                                            &kv->sentinel_lock, &deadline);
             if (w == ETIMEDOUT) break;
         }
         pthread_mutex_unlock(&kv->sentinel_lock);
     } else {
         /* Simple time-box sleep */
         struct timespec req;
         req.tv_sec  = (time_t)timeout_sec;
         req.tv_nsec = (long)((timeout_sec - (double)req.tv_sec) * 1e9);
         nanosleep(&req, NULL);
     }
 
     /* Unsubscribe to stop receiving further messages */
     mosquitto_unsubscribe(kv->mosq, NULL, pattern);
 
     /* Copy results to caller, excluding sentinel topics */
     int out_count = 0;
     pthread_mutex_lock(&kv->entries_lock);
     for (int i = 0; i < kv->entry_count && out_count < max_entries; i++) {
         if (!kv->entries[i].active) continue;
 
         /* Skip sentinel topics */
         pthread_mutex_lock(&kv->sentinel_lock);
         bool is_sent = kvr_is_sentinel(kv, kv->entries[i].topic);
         pthread_mutex_unlock(&kv->sentinel_lock);
         if (is_sent) continue;
 
         out_entries[out_count] = kv->entries[i];
         out_count++;
     }
     pthread_mutex_unlock(&kv->entries_lock);
 
     /* Clear sentinels */
     pthread_mutex_lock(&kv->sentinel_lock);
     kv->sentinel_count = 0;
     pthread_mutex_unlock(&kv->sentinel_lock);
 
     return out_count;
 }
 
 bool kvr_read_single(kvr_store_reader_t *kv,
                      const char *topic,
                      double timeout_sec,
                      char *out_value,
                      int out_value_len)
 {
     kvr_entry_t entries[1];
     int n = kvr_read_pattern(kv, topic, 1, timeout_sec,
                              NULL, false, entries, 1);
     if (n > 0 && strcmp(entries[0].topic, topic) == 0) {
         strncpy(out_value, entries[0].value, out_value_len - 1);
         out_value[out_value_len - 1] = '\0';
         return true;
     }
     return false;
 }
 
 int kvr_read_all(kvr_store_reader_t *kv,
                  const char *base_topic,
                  double timeout_sec,
                  const char *sentinel_topics[],
                  bool wait_for_sentinel,
                  kvr_entry_t *out_entries,
                  int max_entries)
 {
     if (!base_topic) base_topic = "#";
     return kvr_read_pattern(kv, base_topic, 1, timeout_sec,
                             sentinel_topics, wait_for_sentinel,
                             out_entries, max_entries);
 }
 
 bool kvr_is_connected(kvr_store_reader_t *kv)
 {
     pthread_mutex_lock(&kv->state_lock);
     bool c = kv->connected;
     pthread_mutex_unlock(&kv->state_lock);
     return c;
 }

File: zig_programs_and_containers/building_blocks/knowledge_base/mqtt/kv_read_store/src/kv_store_reader.zig
//! kv_store_reader.zig - Reads retained key/value messages from an MQTT broker.
//!
//! Translated from C KVStoreReader.
//! Requires: libmosquitto (apt install libmosquitto-dev)

const std = @import("std");
const log = std.log.scoped(.kv_reader);

const c = @cImport({
    @cInclude("mosquitto.h");
});

// ──────────────────────────────────────────────────────────────────────
//  Constants
// ──────────────────────────────────────────────────────────────────────

pub const MAX_TOPIC_LEN = 256;
pub const MAX_VALUE_LEN = 4096;
pub const MAX_ENTRIES = 256;
pub const MAX_SENTINELS = 8;

// ──────────────────────────────────────────────────────────────────────
//  Types
// ──────────────────────────────────────────────────────────────────────

pub const Config = struct {
    host: [:0]const u8 = "localhost",
    port: u16 = 1883,
    client_id: [:0]const u8 = "kv-reader",
    keepalive: u16 = 60,
    username: ?[:0]const u8 = null,
    password: ?[:0]const u8 = null,
    clean_session: bool = true,
};

pub const Entry = struct {
    topic: [MAX_TOPIC_LEN]u8 = [_]u8{0} ** MAX_TOPIC_LEN,
    value: [MAX_VALUE_LEN]u8 = [_]u8{0} ** MAX_VALUE_LEN,
    active: bool = false,

    /// Return the topic as a slice.
    pub fn topicSlice(self: *const Entry) []const u8 {
        return std.mem.sliceTo(&self.topic, 0);
    }

    /// Return the value as a slice.
    pub fn valueSlice(self: *const Entry) []const u8 {
        return std.mem.sliceTo(&self.value, 0);
    }
};

pub const Error = error{
    MosquittoNew,
    ConnectFailed,
    LoopStartFailed,
    ConnectTimeout,
    NotConnected,
    SubscribeFailed,
};

// ──────────────────────────────────────────────────────────────────────
//  Reader
// ──────────────────────────────────────────────────────────────────────

pub const Reader = struct {
    mosq: *c.struct_mosquitto = undefined,
    cfg: Config,

    // Collected KV entries
    entries: [MAX_ENTRIES]Entry = [_]Entry{.{}} ** MAX_ENTRIES,
    entry_count: usize = 0,
    entries_mutex: std.Thread.Mutex = .{},

    // Sentinel coordination
    sentinels: [MAX_SENTINELS][MAX_TOPIC_LEN]u8 = [_][MAX_TOPIC_LEN]u8{[_]u8{0} ** MAX_TOPIC_LEN} ** MAX_SENTINELS,
    sentinel_count: usize = 0,
    sentinel_fired: bool = false,
    sentinel_mutex: std.Thread.Mutex = .{},
    sentinel_cond: std.Thread.Condition = .{},

    // Subscribe acknowledgement
    sub_acked: bool = false,
    sub_mutex: std.Thread.Mutex = .{},
    sub_cond: std.Thread.Condition = .{},

    // Connection state
    connected: bool = false,
    running: bool = false,
    connect_done: bool = false,
    state_mutex: std.Thread.Mutex = .{},
    connect_cond: std.Thread.Condition = .{},

    const Self = @This();

    // ── Internal helpers ─────────────────────────────────────────────

    /// Check if topic is a sentinel. Caller must hold sentinel_mutex.
    fn isSentinel(self: *Self, topic: []const u8) bool {
        for (0..self.sentinel_count) |i| {
            const sentinel = std.mem.sliceTo(&self.sentinels[i], 0);
            if (std.mem.eql(u8, sentinel, topic)) return true;
        }
        return false;
    }

    /// Store a KV entry. Overwrites if topic exists. Caller must hold entries_mutex.
    fn storeEntry(self: *Self, topic: []const u8, value: []const u8) void {
        // Check for existing — overwrite
        for (0..self.entry_count) |i| {
            if (self.entries[i].active) {
                const existing = std.mem.sliceTo(&self.entries[i].topic, 0);
                if (std.mem.eql(u8, existing, topic)) {
                    const vlen = @min(value.len, MAX_VALUE_LEN - 1);
                    @memcpy(self.entries[i].value[0..vlen], value[0..vlen]);
                    self.entries[i].value[vlen] = 0;
                    return;
                }
            }
        }
        // Append new
        if (self.entry_count < MAX_ENTRIES) {
            var e = &self.entries[self.entry_count];
            const tlen = @min(topic.len, MAX_TOPIC_LEN - 1);
            @memcpy(e.topic[0..tlen], topic[0..tlen]);
            e.topic[tlen] = 0;
            const vlen = @min(value.len, MAX_VALUE_LEN - 1);
            @memcpy(e.value[0..vlen], value[0..vlen]);
            e.value[vlen] = 0;
            e.active = true;
            self.entry_count += 1;
        }
    }

    // ── Public API ───────────────────────────────────────────────────

    pub fn init(cfg: Config) Error!Self {
        var self = Self{ .cfg = cfg };

        const mosq = c.mosquitto_new(cfg.client_id.ptr, cfg.clean_session, null);
        if (mosq == null) {
            log.err("[Error] mosquitto_new failed", .{});
            return Error.MosquittoNew;
        }
        self.mosq = mosq.?;

        if (cfg.username) |user| {
            const pw: ?[*:0]const u8 = if (cfg.password) |p| p.ptr else null;
            _ = c.mosquitto_username_pw_set(self.mosq, user.ptr, pw);
        }

        c.mosquitto_connect_callback_set(self.mosq, onConnectCb);
        c.mosquitto_disconnect_callback_set(self.mosq, onDisconnectCb);
        c.mosquitto_message_callback_set(self.mosq, onMessageCb);
        c.mosquitto_subscribe_callback_set(self.mosq, onSubscribeCb);

        return self;
    }

    pub fn connect(self: *Self, timeout_ms: u64) Error!void {
        c.mosquitto_user_data_set(self.mosq, @ptrCast(self));

        self.state_mutex.lock();
        self.connect_done = false;
        self.running = true;
        self.state_mutex.unlock();

        const rc_conn = c.mosquitto_connect(
            self.mosq,
            self.cfg.host.ptr,
            @intCast(self.cfg.port),
            @intCast(self.cfg.keepalive),
        );
        if (rc_conn != c.MOSQ_ERR_SUCCESS) {
            log.err("[Error] Connection failed: {s}", .{c.mosquitto_strerror(rc_conn)});
            self.running = false;
            return Error.ConnectFailed;
        }

        const rc_loop = c.mosquitto_loop_start(self.mosq);
        if (rc_loop != c.MOSQ_ERR_SUCCESS) {
            log.err("[Error] loop_start failed: {s}", .{c.mosquitto_strerror(rc_loop)});
            self.running = false;
            return Error.LoopStartFailed;
        }

        // Wait for CONNACK
        self.state_mutex.lock();
        defer self.state_mutex.unlock();

        if (!self.connect_done) {
            self.connect_cond.timedWait(&self.state_mutex, timeout_ms * std.time.ns_per_ms) catch {};
        }

        if (!self.connected) {
            log.err("[Error] Connection timeout or refused ({s}:{d})", .{ self.cfg.host, self.cfg.port });
            self.running = false;
            _ = c.mosquitto_loop_stop(self.mosq, true);
            return Error.ConnectTimeout;
        }

        log.info("[Connected] Successfully connected to {s}:{d}", .{ self.cfg.host, self.cfg.port });
    }

    pub fn disconnect(self: *Self) void {
        self.running = false;
        _ = c.mosquitto_disconnect(self.mosq);
        _ = c.mosquitto_loop_stop(self.mosq, false);
        self.state_mutex.lock();
        defer self.state_mutex.unlock();
        self.connected = false;
    }

    pub fn deinit(self: *Self) void {
        if (self.running or self.connected) {
            self.disconnect();
        }
        c.mosquitto_destroy(self.mosq);
    }

    /// Read retained messages matching a topic pattern.
    /// Subscribes, collects retained messages until sentinel fires or timeout.
    /// Returns number of entries written to out_entries.
    pub fn readPattern(
        self: *Self,
        pattern: [:0]const u8,
        qos: c_int,
        timeout_ms: u64,
        sentinel_topics: ?[]const [:0]const u8,
        wait_for_sentinel: bool,
        out_entries: []Entry,
    ) usize {
        if (!self.connected) {
            log.err("[Error] Not connected. Call connect() first.", .{});
            return 0;
        }

        // Clear previous entries
        self.entries_mutex.lock();
        self.entry_count = 0;
        self.entries_mutex.unlock();

        // Setup sentinels
        self.sentinel_mutex.lock();
        self.sentinel_fired = false;
        self.sentinel_count = 0;
        if (sentinel_topics) |sents| {
            for (sents) |s| {
                if (self.sentinel_count >= MAX_SENTINELS) break;
                const slen = @min(s.len, MAX_TOPIC_LEN - 1);
                @memcpy(self.sentinels[self.sentinel_count][0..slen], s[0..slen]);
                self.sentinels[self.sentinel_count][slen] = 0;
                self.sentinel_count += 1;
            }
        }
        self.sentinel_mutex.unlock();

        // Subscribe to pattern
        self.sub_mutex.lock();
        self.sub_acked = false;
        self.sub_mutex.unlock();

        const rc = c.mosquitto_subscribe(self.mosq, null, pattern.ptr, qos);
        if (rc != c.MOSQ_ERR_SUCCESS) {
            log.err("[Error] Subscribe failed for {s}: {s}", .{ pattern, c.mosquitto_strerror(rc) });
            self.sentinel_mutex.lock();
            self.sentinel_count = 0;
            self.sentinel_mutex.unlock();
            return 0;
        }

        // Wait for SUBACK (best-effort, 2s cap)
        {
            self.sub_mutex.lock();
            defer self.sub_mutex.unlock();
            if (!self.sub_acked) {
                self.sub_cond.timedWait(&self.sub_mutex, 2000 * std.time.ns_per_ms) catch {};
            }
        }

        // Wait for retained messages: sentinel or timeout
        if (wait_for_sentinel and self.sentinel_count > 0) {
            self.sentinel_mutex.lock();
            defer self.sentinel_mutex.unlock();
            if (!self.sentinel_fired) {
                self.sentinel_cond.timedWait(&self.sentinel_mutex, timeout_ms * std.time.ns_per_ms) catch {};
            }
        } else {
            std.time.sleep(timeout_ms * std.time.ns_per_ms);
        }

        // Unsubscribe
        _ = c.mosquitto_unsubscribe(self.mosq, null, pattern.ptr);

        // Copy results, excluding sentinels
        var out_count: usize = 0;
        self.entries_mutex.lock();
        defer self.entries_mutex.unlock();

        for (0..self.entry_count) |i| {
            if (!self.entries[i].active) continue;
            if (out_count >= out_entries.len) break;

            const topic = std.mem.sliceTo(&self.entries[i].topic, 0);

            self.sentinel_mutex.lock();
            const is_sent = self.isSentinel(topic);
            self.sentinel_mutex.unlock();
            if (is_sent) continue;

            out_entries[out_count] = self.entries[i];
            out_count += 1;
        }

        // Clear sentinels
        self.sentinel_mutex.lock();
        self.sentinel_count = 0;
        self.sentinel_mutex.unlock();

        return out_count;
    }

    /// Read a single retained value for an exact topic.
    pub fn readSingle(
        self: *Self,
        topic: [:0]const u8,
        timeout_ms: u64,
        out_value: []u8,
    ) bool {
        var entries: [1]Entry = [_]Entry{.{}};
        const n = self.readPattern(topic, 1, timeout_ms, null, false, &entries);
        if (n > 0) {
            const got_topic = std.mem.sliceTo(&entries[0].topic, 0);
            if (std.mem.eql(u8, got_topic, topic)) {
                const val = std.mem.sliceTo(&entries[0].value, 0);
                const clen = @min(val.len, out_value.len - 1);
                @memcpy(out_value[0..clen], val[0..clen]);
                out_value[clen] = 0;
                return true;
            }
        }
        return false;
    }

    /// Read all retained messages under a base topic (default "#").
    pub fn readAll(
        self: *Self,
        base_topic: ?[:0]const u8,
        timeout_ms: u64,
        sentinel_topics: ?[]const [:0]const u8,
        wait_for_sentinel: bool,
        out_entries: []Entry,
    ) usize {
        const pattern = base_topic orelse "#";
        return self.readPattern(pattern, 1, timeout_ms, sentinel_topics, wait_for_sentinel, out_entries);
    }

    pub fn isConnected(self: *Self) bool {
        self.state_mutex.lock();
        defer self.state_mutex.unlock();
        return self.connected;
    }

    // ── Mosquitto C callbacks ────────────────────────────────────────

    fn onConnectCb(_: ?*c.struct_mosquitto, userdata: ?*anyopaque, rc: c_int) callconv(.C) void {
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.state_mutex.lock();
        defer self.state_mutex.unlock();
        self.connected = (rc == 0);
        self.connect_done = true;
        self.connect_cond.signal();
    }

    fn onDisconnectCb(_: ?*c.struct_mosquitto, userdata: ?*anyopaque, _: c_int) callconv(.C) void {
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.state_mutex.lock();
        defer self.state_mutex.unlock();
        self.connected = false;
    }

    fn onMessageCb(_: ?*c.struct_mosquitto, userdata: ?*anyopaque, msg: ?*const c.struct_mosquitto_message) callconv(.C) void {
        const self: *Self = @ptrCast(@alignCast(userdata));
        const m = msg orelse return;

        // Extract topic
        const topic_cstr: [*:0]const u8 = @ptrCast(m.topic orelse return);
        const topic = std.mem.sliceTo(topic_cstr, 0);

        // Extract payload
        var payload_buf: [MAX_VALUE_LEN]u8 = undefined;
        var payload_len: usize = 0;
        if (m.payload != null and m.payloadlen > 0) {
            const pptr: [*]const u8 = @ptrCast(m.payload);
            payload_len = @min(@as(usize, @intCast(m.payloadlen)), MAX_VALUE_LEN - 1);
            @memcpy(payload_buf[0..payload_len], pptr[0..payload_len]);
        }
        payload_buf[payload_len] = 0;
        const payload = payload_buf[0..payload_len];

        // Check sentinel
        self.sentinel_mutex.lock();
        if (self.isSentinel(topic)) {
            self.sentinel_fired = true;
            self.sentinel_cond.signal();
            self.sentinel_mutex.unlock();
            return;
        }
        self.sentinel_mutex.unlock();

        // Normal KV handling
        self.entries_mutex.lock();
        defer self.entries_mutex.unlock();
        if (m.retain) {
            self.storeEntry(topic, payload);
            log.info("[Retained] {s}", .{topic});
        } else {
            log.info("[Non-retained] {s}", .{topic});
        }
    }

    fn onSubscribeCb(_: ?*c.struct_mosquitto, userdata: ?*anyopaque, mid: c_int, _: c_int, _: ?*const c_int) callconv(.C) void {
        const self: *Self = @ptrCast(@alignCast(userdata));
        log.info("[Subscribed] mid={d}", .{mid});
        self.sub_mutex.lock();
        defer self.sub_mutex.unlock();
        self.sub_acked = true;
        self.sub_cond.signal();
    }
};

// ──────────────────────────────────────────────────────────────────────
//  Tests
// ──────────────────────────────────────────────────────────────────────

test "Reader Config defaults" {
    const cfg = Config{};
    try std.testing.expectEqualStrings("localhost", cfg.host);
    try std.testing.expectEqual(@as(u16, 1883), cfg.port);
    try std.testing.expectEqualStrings("kv-reader", cfg.client_id);
    try std.testing.expect(cfg.clean_session);
}

test "Entry sliceTo" {
    var entry = Entry{};
    const src = "test/topic";
    @memcpy(entry.topic[0..src.len], src);
    entry.topic[src.len] = 0;
    entry.active = true;
    try std.testing.expectEqualStrings("test/topic", entry.topicSlice());
}